src/context.rs

```diff
@@ -1789,19 +1836,43 @@ impl Context {
             Error::local_error(ErrorKind::MissingAuthSession)
         })
     }
+
+    /// Function that returns the handle as an ESYS_TR if it
+    /// was valid or an error if it was None.
+    fn ensure_not_none(handle: &ObjectHandle, fn_name: &str) -> Result<ESYS_TR> {
+        if *handle != ObjectHandle::None {
+            Ok((*handle).into())
+        } else {
+            error!("ObjectHandle::None cannot be used in call to {}", fn_name);
+            Err(Error::local_error(ErrorKind::InvalidParam))
+        }
+    }
 }
 
 impl Drop for Context {
     fn drop(&mut self) {
         info!("Closing context.");
 
-        // Flush the open handles.
-        self.open_handles.clone().iter().for_each(|handle| {
-            info!("Flushing handle {}", ESYS_TR::from(*handle));
-            if let Err(e) = self.flush_context(*handle) {
+        // Flush handles
+        for handle in self.handle_manager.handles_to_flush() {
+            info!("Flushing handle {}", ESYS_TR::from(handle));
+            if let Err(e) = self.flush_context(handle) {
                 error!("Error when dropping the context: {}.", e);
             }
-        });
+        }
+
+        // Close handles
+        for handle in self.handle_manager.handles_to_close().iter_mut() {
+            info!("Closing handle {}", ESYS_TR::from(*handle));
+            if let Err(e) = self.tr_close(handle) {
+                error!("Error when dropping context: {}.", e);
+            }
+        }
+
+        // Check if all handles have been cleaned up proeprly.
+        if self.handle_manager.has_open_handles() {
+            error!("Not all handles have had its rersources successfully released");
+        }
```

Comment:
I guess it's a sanity check, but is this necessary? We have a mutable reference to `self`, so nothing could've been added between those calls above which together cover everything in the `HashMap`

Reply:
It is a "This should never happen" kind of check.